transport/http/roundtrip_test.go

```diff
@@ -0,0 +1,161 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package http
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/internal/testtime"
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+func TestRoundTripSuccess(t *testing.T) {
+	headerKey, headerVal := "foo", "bar"
+	giveBody := "successful response"
+
+	echoServer := httptest.NewServer(http.HandlerFunc(
+		func(w http.ResponseWriter, req *http.Request) {
+			defer req.Body.Close()
+
+			// copy header
+			header := req.Header.Get(headerKey)
+			w.Header().Set(headerKey, header)
+
+			// copy body
+			body, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				t.Error("error reading body")
+			}
+			_, err = w.Write(body)
+			if err != nil {
+				t.Error("error writing body")
+			}
+		},
+	))
+	defer echoServer.Close()
+
+	// start outbound
+	httpTransport := NewTransport()
+	out := httpTransport.NewSingleOutbound(echoServer.URL)
+	require.NoError(t, out.Start(), "failed to start outbound")
+	defer out.Stop()
+
+	// create request
+	hreq, err := http.NewRequest("GET", echoServer.URL, bytes.NewReader([]byte(giveBody)))
+	require.NoError(t, err, "could not create HTTP request")
+	hreq.Header.Add(headerKey, headerVal)
+
+	// add deadline
+	ctx, cancel := context.WithTimeout(context.Background(), testtime.Second)
+	defer cancel()
+	hreq = hreq.WithContext(ctx)
+
+	// make call
+	client := http.Client{Transport: out}
+	res, err := client.Do(hreq)
+	require.NoError(t, err, "could not make call")
+	defer res.Body.Close()
+
+	// validate header
+	gotHeaderVal := res.Header.Get(headerKey)
+	assert.Equal(t, headerVal, gotHeaderVal, "header did not match")
+
+	// validate body
+	gotBody, err := ioutil.ReadAll(res.Body)
+	require.NoError(t, err)
+	assert.Equal(t, giveBody, string(gotBody), "body did not match")
+}
+
+func TestRoundTripTimeout(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(
+		func(w http.ResponseWriter, r *http.Request) {
+			time.Sleep(time.Hour) // never respond
+		}))
+
+	// start outbound
+	out := NewTransport().NewSingleOutbound(server.URL)
+	require.NoError(t, out.Start(), "failed to start outbound")
+	defer out.Stop()
+
+	// create request
+	req, err := http.NewRequest("POST", server.URL, nil /* body */)
+	require.NoError(t, err)
+
+	// set a small deadline so the the call times out quickly
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond)
+	defer cancel()
+	req = req.WithContext(ctx)
+
+	// make call
+	client := http.Client{Transport: out}
+	res, err := client.Do(req)
+
+	// validate response
+	if assert.Error(t, err) {
+		assert.Contains(t, err.Error(), "deadline-exceeded")
+	}
+	if assert.NotNil(t, ctx.Err()) {
+		assert.Equal(t, "context deadline exceeded", ctx.Err().Error())
```

Comment:
The first one should probably use [`IsDeadlineExceeded`](https://godoc.org/go.uber.org/yarpc/yarpcerrors#IsDeadlineExceeded) and the second an [`== context.DeadlineExceeded`](https://golang.org/pkg/context/#DeadlineExceeded).

Reply:
Can't use `yarpcerrors.IsDeadlineExceeded` since the returned error is really a `url.Error`(https://golang.org/pkg/net/url/#Error). I've updated the test with a comment and used the `String` method on the error code instead.